IPs/RSKIP244.md

```diff
@@ -0,0 +1,125 @@
+
+> The value corresponding to a mapping key k is located at keccak256(h(k) . p) where "." is concatenation and h is a function that is applied to the key depending on its type:  (note: p is considered a uint256)
+>
+> * For value types, h pads the value to 32 bytes in the same way as when storing the value in memory.
+>
+> - For strings and byte arrays, `h` computes the `keccak256` hash of the unpadded data.
+>
+>   
+>
+> If the value is again a non-elementary type, the positions are found by adding an offset of keccak256(k . p).
+
+Solidity uses Keccak256(..) for storage keys because it uses adjacent cells when the value in the mapping occupies more than 32 bytes. It can't use simply (k. p), because an attacker could choose the key k=(u+1) for a pre-existing key u, in order to store a certain value overlapping the next 32 bytes of the data stored at key u. 
+But because ERC20 contract store uint256 balances (one cell),  the balance mapping doesn't need this protection as long as the address space does not collude with the address space of a Solidity native contract field.
+Instead of using Solidity mappings, the program can use SSTORE/SLOAD with a user-defined key mapping such as: `<addr20> 0xff`. This mapping does not collude with Solidity mappings because Solidity mapping storage keys are always the hash digest of 64 bytes of data, and the proposed pre-image is only 21 bytes in length. However, it can collude with a field variable at slot 0xff if addr20 is all zeros. The reason is that the EVM always expands the keys to 256 bits RSK, filling the MSBs with zero. For such as contract field to exists the programmer must define 255 contract fields, which is uncommon. We could avoid potential problem this by adding a non-zero byte before addr20 like this: `0xff <addr20> 0xff`. If we ditch the trailing 0xff, then it can't collude also. So a final proposal is simply: `0xff addr20`. 
+
+By using the proposed scheme to store an ERC20 balance the platform saves:
+
+* 32 bytes for the trie node hash
+* 11 bytes of storage cell key
+* 2 bytes of a node flags and path length
+* At least 32 bytes of LevelDB overhead 
+
+The total saving is 32+11+2+32=77 bytes. When stored embedded in a node the space used by a storage cell is: 20 bytes (key) + 12 bytes (value) + 10 bytes (hashed key prefix) + 2 bytes (node flags and path length) = 44 bytes. When not embedded the space required is 44 + 77 = 121 bytes.
+Therefore, the cost for using small nodes for balances can be as low as 36% of the current cost. 
+
+To conclude, a transfer of ERC20 tokens currently cost approximately 10K gas (with pre-existing non-zero balances). And optimized ERC20 will only reduce this cost to 9420 gas. When transferring to addresses without previous balance, the current scheme costs 25K, while the proposed scheme costs ~13K gas.
```

Comment:
Clarify that you are only talking about costs related to storage .. not overall cost of token transfer